### DISCUSSION TITLE
Face recognition for multi user

### DISCUSSION THREAD

**Discussion by Ulrar** [toverainc/willow#86] [Ideas]
Hi,

There's mentions of multi users, just an idea on one of the ways this may work is face recognition. There was a post on reddit recently about someone who managed to get that working using a cheap (relatively speaking) mmwave sensor on an ESP32, not sure how heavy that is and if it could run at the same time as Willow but that could be one use for the GPIO pins.

The thread in question : https://www.reddit.com/r/homeassistant/comments/13mru0s/facerec_beta_simple_fast_local_and_secure_face/

Feel free to close this if it doesn't align with the vision, just a thought :)

**Comment by kristiankielhofner** [comment]
Some issues as I see them:

- We're a far field audio device with wake word. Using anything based on a camera requires close and direct line of sight. If you're going to be right in front of the Willow device staring at it you may as well tap a button :).
- There are always demos around for projects online. In many cases the real-world experience is very different... Without criticizing or questioning this project specifically we don't do any staging, upload the one that worked, etc with Willow. We understand what it's like to be an early project (we are one ourselves) but in many cases the real-world experience for things such as this are drastically different than the one take uploaded by the creators.
- It takes time. This project starts recognition every three seconds, which is roughly 5-6x our response time to actually complete the action via Home Assistant with the detected command. I just tried it quickly and in three seconds I can take my phone out of my pocket and Face ID unlocks. The HA dashboard is on my home screen and takes me roughly one additional second to open. In essence, this is cool but it's faster and many times more reliable to just take your phone out of your pocket and use the HA dashboard at this point.
- Wake needs to be reliable. I always say "if you don't have wake, you don't have anything" which is to say it's impossible to interface with anything via speech if the device doesn't wake reliably.

Anything supporting multi-user (either via audio or visually) is inherently going to drastically reduce the user experience of interacting with Willow and the platforms it supports. Not to say that we'll never do it via some means but generally speaking I also like to say that we focus on real-world utility, usability, practicality - not cool demos/party tricks. The Willow Inference Server could potentially be extended to detect multiple users via voice eventually but even that would have to be *REALLY* solid for us to include it.

That said, of course Willow is open source and I would actually be entertained by people adding functionality such as this, but in many/most cases they likely wouldn't make it in tree.
